pandas/core/algorithms.py

```diff
@@ -631,6 +635,11 @@ def factorize(
             uniques, codes, na_sentinel=na_sentinel, assume_unique=True, verify=False
         )
 
+    code_is_na = codes == na_sentinel
+    if not dropna and code_is_na.any():
+        uniques = np.append(uniques, [None])
```

Comment:
this is likley changing the dtype; you need the na for this type, use pandas.core.dtypes.cast.na_value_for_dtype.

Reply:
i do not think so, first of all, if this changed using `cast.na_value_for_dtype`, then `_reconstruct_data` will fail since it cannot cast `NaN` to `integer`, and secondly this `_reconstruct_data` will cast the data to the correct type, so as you see from the output, all data are casted correctly.

Reply:
Is leveraging the Int64 type an option here? This just seems a little strange in its current state